languages/rust/oso/src/host/from_polar.rs

```diff
@@ -135,6 +151,76 @@ impl<T: FromPolar> FromPolar for Vec<T> {
     }
 }
 
+impl<T: FromPolar> FromPolar for LinkedList<T> {
+    fn from_polar(val: PolarValue) -> crate::Result<Self> {
+        if let PolarValue::List(l) = val {
+            let mut result = LinkedList::new();
+            for v in l {
+                result.push_back(T::from_polar(v)?);
+            }
+            Ok(result)
+        } else {
+            Err(TypeError::expected("List").user())
+        }
+    }
+}
+
+impl<T: FromPolar> FromPolar for VecDeque<T> {
+    fn from_polar(val: PolarValue) -> crate::Result<Self> {
+        if let PolarValue::List(l) = val {
+            let mut result = VecDeque::new();
+            for v in l {
+                result.push_back(T::from_polar(v)?);
+            }
+            Ok(result)
+        } else {
+            Err(TypeError::expected("List").user())
+        }
+    }
+}
+
+impl<T: Eq + Hash + FromPolar> FromPolar for HashSet<T> {
+    fn from_polar(val: PolarValue) -> crate::Result<Self> {
+        if let PolarValue::List(l) = val {
+            let mut result = HashSet::new();
+            for v in l {
+                result.insert(T::from_polar(v)?);
+            }
+            Ok(result)
+        } else {
+            Err(TypeError::expected("List").user())
+        }
+    }
+}
+
+impl<T: Eq + Ord + FromPolar> FromPolar for BTreeSet<T> {
```

Comment:
What is the `Eq` bound needed for?

Reply:
Because a set cannot have two equal elements like a Map can.

Also note how the [declaration for BTreeSet](https://doc.rust-lang.org/src/alloc/collections/btree/set.rs.html#62-66) imples the need for `Eq` and `Ord` while the [declaration for BTreeMap](https://doc.rust-lang.org/src/alloc/collections/btree/map.rs.html#132-136) does not.

Reply:
Ah it's derived on struct itself! That's why I couldn't see it in the docs. Thanks for clarifying!